feat: Allow metavariable-pattern to be more widely used with generic mode

nmote opened the pull request:
As I've explained inline in more detail, we now allow metavariable-pattern to be used on any kind of metavariable binding as long as the target language for the metavariable-pattern is generic. Previously we only allowed metavariable-pattern to be used with a sub language if the metavariable in question was bound to a string literal.

Test plan: Automated tests

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

nmote commented:
Don't merge -- I'll eat the conflicts with https://github.com/returntocorp/semgrep/pull/7052

IagoAbal commented on src/engine/Metavariable_pattern.ml:
```diff
@@ -118,53 +118,79 @@ let satisfies_metavar_pattern_condition nested_formula_has_matches env r mvar
+                 *
+                 * However, now we also allow non-text metavariable bindings to
+                 * match in generic mode. Because non-text metavariable bindings
+                 * are highly unlikely to be syntactically valid in any
```
Do tests break if you allow this for any language and not just `generic` ?

nmote replied:
No. The one test that broke was updated in https://github.com/returntocorp/semgrep-rules/pull/2691. @LewisArdern agreed that the new behavior here makes sense, and updated the rule accordingly.

I wouldn't object strongly to allowing this for any language, but I would want to see a motivating use case.

IagoAbal commented on src/engine/Metavariable_pattern.ml:
```diff
@@ -118,53 +118,79 @@ let satisfies_metavar_pattern_condition nested_formula_has_matches env r mvar
+              | None ->
+                  (* This is not necessarily an error in the rule, e.g. you may be
+                   * matching `$STRING + ...` and then add a metavariable-pattern on
+                   * `$STRING`. This will only work when `$STRING` binds to some text
```
~Need to update this comment?~ Nah it's still valid, it just doesn't apply to `generic` anymore.

nmote replied:
I've added a clarification.

IagoAbal commented on changelog.d/metavar-pattern-generic.feat:
```diff
@@ -0,0 +1 @@
+Allow metavariable-pattern clauses that use `language: generic` to run (and potentially match) on any metavariable binding kind, not just strings.
```
Isn't this change associated with a customer-related ticket? I would expect the ticker number to be used for the changelog entry instead of `metavar-pattern-generic.feat` and referenced in the PR comment as well.

IagoAbal replied:
We could also add more detail in this changelog entry (like an example use case), so that tech-writers can highlight it appropriately in the release notes. cc @adamkvitek

nmote replied:
Well, there are going to be at least 3 PRs associated with that ticket and they can't all use the same changelog file. Doing it this way seems to make the most sense to me.

nmote replied:
I've added an example to the changelog entry.

IagoAbal commented:
Nice addition!

nmote committed:
feat: Allow metavariable-pattern to be more widely used with generic mode

As I've explained inline in more detail, we now allow
metavariable-pattern to be used on any kind of metavariable binding as
long as the target language for the metavariable-pattern is generic.
Previously we only allowed metavariable-pattern to be used with a sub
language if the metavariable in question was bound to a string literal.

Test plan: Automated tests
